### ISSUE TITLE
[Enhancement] Song Time and External API

### ISSUE THREAD

**Issue by Coder-256** [beardedspice/beardedspice#332]
Hi,

I propose adding two optional methods to each media strategy: `getTime` and (maybe) `setTime`.

`getTime` should return the current playing time in seconds (ex. if I'm 2 minutes and 12 seconds into a song, it should return 132 seconds).

`setTime` should set the current time to however many seconds specified in its input (ex. fast forward or reverse the current song to 2 minutes and 32 seconds for 132 second input).

In addition, I propose adding some sort of external API. Here are some examples:

- A CLI
- An [Applescript Dictionary](https://developer.apple.com/library/mac/documentation/AppleScript/Conceptual/AppleScriptX/Concepts/scriptable_apps.html) (probably the easiest to code _and_ to use)
- Automator Actions

This would allow users to use information such as song title, time (^^^), etc. in their apps.
For example, I was thinking of making a lyrics app. All (the two) Mac lyrics apps seem to have either disappeared or stopped working ([Musixmatch](https://www.musixmatch.com/) and [MiniLyrics](http://crintsoft.com/MiniLyrics.php), respectively). I realized that it would be a huge pain to manually write my own code for accessing all these different websites in all these different browsers, so it would be much easier to have an API to access this info through the already created program BeardedSpice.

One problem - this would make BeardedSpice a dependency for my app, and (crazy) users may not want it installed. So, while this may sound pretty ambicious, I think that an API should be made for accessing websites and playing/pausing, etc. and the BeardedSpice app should be separated and rewritten using that API. While at first this may seem crazy, it actually won't require much rewriting - more just separating files and documenting the API (and better documenting how to write strategies for it, just saying as the creator of the Udemy MediaStrategy).

Thank you for considering this! I truly believe this will lead to a better experience for developers and, subsequently, end users.

**Comment by Stillness-2** [comment]
Good idea. Will be interesting if all logic of "players" will be separate from AppDelegate.m into "one high level module".
Go for it. :) Jot down a list of objects/methods that should be in your opinion.

**Comment by Coder-256** [comment]
@Stillness-2 That isn't actually what I meant, but is also a good idea. So, to sum it all up, this is really three different potential enhancements (in order of hardest to easiest to implement):

1. Switch from hard-coded Objective-C strategies to maybe even text fields in the application settings or similar that allow the user to modify each strategy, or maybe even create their own without knowing Objective-C. Users could choose from regex or Javascript to determine if each tab is accepted, simply paste Javascript code into text fields for the different methods, not need to download XCode, and make simpler pull requests. If you want to go all out on this, you could even provide the option to run Applescript or JXA instead of normal Javascript, and even convert the native tab adapters to this new system.
2. Add `setTime` and `getTime` method stubs and a new `Track.length` property for `MediaStrategies`. `setTime` fast forwards or rewinds the current track to the time provided in milliseconds. `getTime` gets the current time of the track in milliseconds. `Track.length` optionally may be set to the length of the track. Here's an example for Google Music:

    ```
    -(NSNumber *)getTime:(TabAdapter *)tab
    {
        NSNumber *value = [tab executeJavascript:@"(function(){return 
    document.querySelector(\"#material-player-progress\").attributes[\"aria-valuenow\"].value;}())"];
        return value;
    }
    //...
    track.length = [tab executeJavascript:@"(function(){return document.querySelector(\"#material-player-progress\").attributes[\"aria-valuemax\"].value;}())"];
    // setTime cannot be implemented for Google Music, and is probably the hardest to implement in general, but should still be included.
    ```
Attempts have already been made. See the Audible `Track.artist`.

3. Create some sort of way for any application to access tabs through BeardedSpice. Here's an example showing what I mean: ![beardedspiceapi](https://cloud.githubusercontent.com/assets/10134823/13050726/49c09d34-d3c3-11e5-8007-6203582bf77c.jpg)
As I said in my first post, a few examples of how this could be implemented are A CLI, [an Applescript Dictionary](https://developer.apple.com/library/mac/documentation/AppleScript/Conceptual/AppleScriptX/Concepts/scriptable_apps.html) (probably the easiest to code and to use), or Automator Actions.

**Comment by Porco-Rosso** [comment]
I also support an external api.
For instance, I wish I could query beardedspice for the current track to display it with [uebersicht](https://github.com/felixhageloh/uebersicht)

**Comment by Coder-256** [comment]
@Porco-Rosso Thanks for the example. For example, as mentioned in my first post, I was thinking of making a lyrics app using `trackInfo`. Any other examples, anybody?

**Comment by Stillness-2** [comment]
> "Switch from hard-coded Objective-C strategies"
For the average person there is no problem to write a new strategy in Objective C. Example of the strategy in the readme.md simple and self explanatory. :)
Again, You didn't stop the fact, that the project on Objective C :)

**Comment by Coder-256** [comment]
A good solution for that is #302. It's not ideal, but it's progress.

**Comment by Stillness-2** [comment]
Did you read discussion in #302? I've stated my position there. :)
I don't see the value in replacing the way of programming strategy! 
Importance only appears when creating/editing strategy will not require changes in the project repository.

**Comment by Coder-256** [comment]
It would just make it easier to develop and update, and I see no cons (all you need to do is accept the PR). So basically, why not? I care most about the external API, anyway.

**Comment by Stillness-2** [comment]
"all you need to do is accept the PR"
Why do you say that? If so to reason, I can to do so now. Here is a recent example of PR #311 (#337 also). In this PR error in javascript. Doesn't matter what language written strategy, because if I just accept PR, without checking, I dramatically worsen the quality of the output product.

 I probably do not understand something about open source, and you right. :)

About external API, i like the idea to do it using AppleScript.

**Comment by Porco-Rosso** [comment]
@Stillness-2  Is an external api still being considered?

**Comment by Stillness-2** [comment]
@Porco-Rosso Yes

**Comment by grounded042** [comment]
Any update on the external API? I would also like to use it with uebersicht.

**Comment by 1ps0** [comment]
It's a long ways off. It needs to be properly spec'd and evaluated first, with at least one example application to demonstrate the roadmap. I imagine it'd become a secondary focus of the app entirely, so we have to consider that we might just be adding feature creep.

What we can look closer at doing in the near future is enabling more applescript paths, which should help you on the side. If you look at #516 you can see how some other people are currently using our limited existing support.

I'll try to put together a formal ticket and add it as a milestone.  In the mean time if you want to write out in detail here what your use case is, it would help with moving forward a lot :)

**Comment by grounded042** [comment]
@1ps0 thanks! #516 was actually a huge help and mostly got me what I wanted. 

What I'm looking for is a simple read-only API. Something that I can query to get:

- track info: title, album, artist, track length
- possibly a url to album art
- current track play timestamp

AppleScript works, but not super well.

**Comment by Porco-Rosso** [comment]
Hi, as I posted in the following thread: https://github.com/beardedspice/beardedspice/issues/516
The applescript stopgap has stopped working, as of now only the player in action is output. 

Is there another variable to call or another way to get the track and artist fields?

**Comment by alice0meta** [comment]
i would use beardedspice if it had a good api

**Comment by 1ps0** [comment]
@alice0meta Can you be more specific? The better detail, the better implementation.
